In [1]:
from tensorflow.keras import datasets
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Flatten, Dropout,Dense

In [2]:
!nvidia-smi

Thu Oct  2 12:18:50 2025       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 550.54.15              Driver Version: 550.54.15      CUDA Version: 12.4     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  Tesla T4                       Off |   00000000:00:04.0 Off |                    0 |
| N/A   47C    P8             11W /   70W |       0MiB /  15360MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+-----

Unduh dan siapkan dataset yang digunakan untuk pelatihan model. Dalam hal ini sama dengan dataset pada eksperimen sebelumnya:



## Unduh Dataset

In [3]:
fashion = datasets.fashion_mnist

(X_train, y_train), (X_test, y_test) = fashion.load_data()

29515/29515 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
26421880/26421880 ━━━━━━━━━━━━━━━━━━━━ 2s 0us/step
5148/5148 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
4422102/4422102 ━━━━━━━━━━━━━━━━━━━━ 1s 0us/step


Lakukan normalisasi nilai piksel gambar sehingga bernilai 0 – 1:



In [4]:
X_train = X_train/255.0
X_test  = X_test/255.0

## Arsitektur Model Pertama

ANN dengan 1 Hidden Layer dan SGD Optimizer

Siapkan model ANN dengan 1 hidden layer yang memiliki 128 node:

In [6]:
model1 = Sequential()
model1.add(Flatten(input_shape=[28, 28]))
model1.add(Dense(128, activation='relu'))
model1.add(Dense(10, activation='softmax'))

/usr/local/lib/python3.12/dist-packages/keras/src/layers/reshaping/flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Lakukan kompilasi model menggunakan SGD sebagai optimizer:



In [7]:
model1.compile(loss='sparse_categorical_crossentropy',
optimizer='SGD',
metrics=['accuracy'])

In [8]:
model1.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ flatten (Flatten)               │ (None, 784)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 128)            │       100,480 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 10)             │         1,290 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 101,770 (397.54 KB)

 Trainable params: 101,770 (397.54 KB)

 Non-trainable params: 0 (0.00 B)

Terdapat total 101,770 paramater yang akan dilatih yang sebagian besar berada pada hidden layer.

Jalankan eksperimen sampai 100 epochs dengan 20% dari Training Set dijadikan sebagai Validation Set.

In [9]:
model1.fit(X_train, y_train, epochs=100,
validation_split=0.2,
batch_size=128)

Epoch 1/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 3s 5ms/step - accuracy: 0.5295 - loss: 1.5343 - val_accuracy: 0.7378 - val_loss: 0.8069
Epoch 2/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.7474 - loss: 0.7689 - val_accuracy: 0.7814 - val_loss: 0.6682
Epoch 3/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.7860 - loss: 0.6513 - val_accuracy: 0.7965 - val_loss: 0.6067
Epoch 4/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8031 - loss: 0.5983 - val_accuracy: 0.8081 - val_loss: 0.5684
Epoch 5/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8130 - loss: 0.5611 - val_accuracy: 0.8160 - val_loss: 0.5439
Epoch 6/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8222 - loss: 0.5356 - val_accuracy: 0.8157 - val_loss: 0.5329
Epoch 7/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8292 - loss: 0.5150 - val_accuracy: 0.8226 - val_loss: 0.5137
Epoch 8/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8276 - loss: 0.5081 - val_accu

Lakukan evaluasi model dengan menggunakan Test Set:


In [10]:
model1.evaluate(X_test, y_test)


313/313 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step - accuracy: 0.8698 - loss: 0.3705


[0.3752034902572632, 0.866599977016449]

## Arstiektur Model Kedua

DNN dengan 2 Hidden Layer dan SGD Optimizer

Siapkan model DNN dengan 2 hidden layer yang masing-masing layer terdiri dari 128 node:

In [12]:
model2 = Sequential()
model2.add(Flatten(input_shape=[28, 28]))
model2.add(Dense(128, activation='relu'))
model2.add(Dense(128, activation='relu'))
model2.add(Dense(10, activation='softmax'))

In [13]:
model2.compile(loss='sparse_categorical_crossentropy',
optimizer='SGD', metrics=['accuracy'])

In [14]:
model2.summary()

Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ flatten_1 (Flatten)             │ (None, 784)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 128)            │       100,480 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_3 (Dense)                 │ (None, 128)            │        16,512 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_4 (Dense)                 │ (None, 10)             │         1,290 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 118,282 (462.04 KB)

 Trainable params: 118,282 (462.04 KB)

 Non-trainable params: 0 (0.00 B)

Terdapat total 118,282 paramater yang akan dilatih dan sebagian besar berada pada hidden layer. Jumlah paramater ini sedikit lebih besar dibanding dengan model1.

Jalankan eksperimen sampai 100 epochs dengan 20% dari Training Set dijadikan sebagai Validation Set:

In [15]:
model2.fit(X_train, y_train, epochs=100, validation_split=0.2, batch_size=128)

Epoch 1/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 3s 5ms/step - accuracy: 0.5160 - loss: 1.5511 - val_accuracy: 0.7175 - val_loss: 0.8013
Epoch 2/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.7407 - loss: 0.7562 - val_accuracy: 0.7654 - val_loss: 0.6616
Epoch 3/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.7851 - loss: 0.6325 - val_accuracy: 0.7941 - val_loss: 0.5920
Epoch 4/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8092 - loss: 0.5715 - val_accuracy: 0.8053 - val_loss: 0.5585
Epoch 5/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8184 - loss: 0.5350 - val_accuracy: 0.8107 - val_loss: 0.5340
Epoch 6/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8247 - loss: 0.5104 - val_accuracy: 0.8248 - val_loss: 0.5020
Epoch 7/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8355 - loss: 0.4884 - val_accuracy: 0.8312 - val_loss: 0.4845
Epoch 8/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8388 - loss: 0.4719 - val_accu

In [16]:
model2.evaluate(X_test, y_test)

313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8769 - loss: 0.3494


[0.35648590326309204, 0.8737000226974487]

## Arsitektur Model Ketiga

DNN dengan 2 Hidden Layer dengan Adam Optimizer

Siapkan model DNN dengan 2 hidden layer yang masing-masing layer terdiri dari 128 node. Arsitektur model ini persis sama dengan arsitektur model2:

In [17]:
model3 = Sequential()
model3.add(Flatten(input_shape=[28, 28]))
model3.add(Dense(128, activation='relu'))
model3.add(Dense(128, activation='relu'))
model3.add(Dense(10, activation='softmax'))

In [18]:
# Berbeda dengan model2, model3 menggunakan Adam optimizer sebagai ganti SGD:

model3.compile(loss='sparse_categorical_crossentropy',
optimizer='Adam',
metrics=['accuracy'])

In [19]:
model3.summary()

Model: "sequential_2"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ flatten_2 (Flatten)             │ (None, 784)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_5 (Dense)                 │ (None, 128)            │       100,480 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_6 (Dense)                 │ (None, 128)            │        16,512 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_7 (Dense)                 │ (None, 10)             │         1,290 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 118,282 (462.04 KB)

 Trainable params: 118,282 (462.04 KB)

 Non-trainable params: 0 (0.00 B)

In [20]:
model3.fit(X_train, y_train, epochs=100,
validation_split=0.2,
batch_size=128)

Epoch 1/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 4s 6ms/step - accuracy: 0.7213 - loss: 0.7979 - val_accuracy: 0.8476 - val_loss: 0.4281
Epoch 2/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8513 - loss: 0.4163 - val_accuracy: 0.8573 - val_loss: 0.3896
Epoch 3/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8716 - loss: 0.3573 - val_accuracy: 0.8561 - val_loss: 0.4015
Epoch 4/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8744 - loss: 0.3348 - val_accuracy: 0.8737 - val_loss: 0.3450
Epoch 5/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8844 - loss: 0.3099 - val_accuracy: 0.8673 - val_loss: 0.3530
Epoch 6/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8913 - loss: 0.2920 - val_accuracy: 0.8733 - val_loss: 0.3498
Epoch 7/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8958 - loss: 0.2810 - val_accuracy: 0.8833 - val_loss: 0.3235
Epoch 8/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8994 - loss: 0.2684 - val_accu

In [21]:
model3.evaluate(X_test, y_test)

313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8834 - loss: 0.9186


[0.926471471786499, 0.8813999891281128]

## Arsitektur Model Keempat

DNN dengan 2 Hidden Layer dengan Adam Optimizer dan Dropout

Siapkan model DNN dengan 2 hidden layer yang masing-masing layer terdiri dari 128 node, serta tambahkan Dropout layer masing-masing setelah input dan hidden layer:

In [22]:
model4 = Sequential()
model4.add(Flatten(input_shape=[28, 28]))
model4.add(Dropout(0.3))
model4.add(Dense(128, activation='relu'))
model4.add(Dropout(0.3))
model4.add(Dense(128, activation='relu'))
model4.add(Dropout(0.3))
model4.add(Dense(10, activation='softmax'))

In [23]:
# Perhatikan, kita tidak menambahkan Dropout setelah output layer. Lakukan kompilasi model menggunakan Adam sebagai optimizer:

model4.compile(loss='sparse_categorical_crossentropy',
optimizer='Adam',
metrics=['accuracy'])

In [24]:
model4.summary()

Model: "sequential_3"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ flatten_3 (Flatten)             │ (None, 784)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout (Dropout)               │ (None, 784)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_8 (Dense)                 │ (None, 128)            │       100,480 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_1 (Dropout)             │ (None, 128)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_9 (Dense)                 │ (None, 128)            │        16,512 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_2 (Dropout)             │ (None, 128)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_10 (Dense)                │ (None, 10)             │         1,290 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 118,282 (462.04 KB)

 Trainable params: 118,282 (462.04 KB)

 Non-trainable params: 0 (0.00 B)

In [25]:
# Jalankan eksperimen sampai 100 epochs dengan 20% dari Training Set dijadikan sebagai Validation Set:

model4.fit(X_train, y_train, epochs=100,
validation_split=0.2,
batch_size=128)

Epoch 1/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step - accuracy: 0.6101 - loss: 1.0680 - val_accuracy: 0.8265 - val_loss: 0.4758
Epoch 2/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.7967 - loss: 0.5636 - val_accuracy: 0.8438 - val_loss: 0.4396
Epoch 3/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8171 - loss: 0.5013 - val_accuracy: 0.8467 - val_loss: 0.4110
Epoch 4/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8289 - loss: 0.4714 - val_accuracy: 0.8608 - val_loss: 0.3855
Epoch 5/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8317 - loss: 0.4547 - val_accuracy: 0.8606 - val_loss: 0.3757
Epoch 6/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8364 - loss: 0.4434 - val_accuracy: 0.8648 - val_loss: 0.3703
Epoch 7/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8459 - loss: 0.4263 - val_accuracy: 0.8674 - val_loss: 0.3646
Epoch 8/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step - accuracy: 0.8456 - loss: 0.4157 - val_accu

In [26]:
model4.evaluate(X_test, y_test)

313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8889 - loss: 0.3060


[0.3101238012313843, 0.887499988079071]